### DISCUSSION TITLE
Generating test vector in Sionna

### DISCUSSION THREAD

**Discussion by cacent96** [NVlabs/sionna#62] [Q&A] [answered]
Hello,

Is it possible to generate the test vector using Sionna? This test vector will be loaded to USRP and transmit. I'm trying to create a new module and generating the corresponding test vector. If yes, then how do I do it??

Many thanks!!

**Comment by SebastianCa** [comment, marked as answer]
In principle, you can export any signal/tensor generated with Sionna and load it as test vector for your USRP setup. 

Could you please elaborate further on what kind of test vector you want to generate?

**Reply by cacent96** [reply]
For example, I will create a new module for the new 6G candidates waveform lets say OTFS or OFDM for now, and this will be input to the transmitter block. Considering the we have one BS and multiple UE case, how can the generation of vector be done in Sionna?

**Reply by SebastianCa** [reply]
You can find a detailed example of MU-MIMO simulations with Sionna in the [example notebook](https://nvlabs.github.io/sionna/examples/Realistic_Multiuser_MIMO_Simulations.html). In your case you would simply replace the channel with the USRP transmission.

Without knowing the exact details of your project, I would simply run the waveform generating module in Sionna (in your case the BS) and save the generated signal as numpy array. Hint: `<tensor_name>.numy()` returns the values of a tensor as numpy array.

After transmission, you can import the recorded data-sequence in Sionna and run the receiver functions (in your case the UE).  Hint: `tf.constant(<numpy_array>, dt=tf.float32)` converts your numpy array into a tensor.

**Reply by cacent96** [reply]
Thank you for your reply. I have another question,

Let's say I have one signal vector generated from Sionna and transmitted via USRP. In Multi-user case, the received signal obtained from the received USRP will be processed in the Sionna, by applying different channel effect to multi-user and process them in Parallel using GPU. This mean we are applying the channel module at the receiver side. 

Question is,
1) Can this be done in Sionna?
2) In real-time, near real-time, or offline processing?

Thank you

**Reply by SebastianCa** [reply]
I’m not sure if I fully understand your exact setup, however, if the channel is replaced by USRPs/data, there is no need to additionally apply the channel module. Note that: in case you want to do multi-user experiments with actual over-the-air data, you would need multiple USRP receivers and carefully set-up your over-the-air experiment (e.g., consider synchronization,…).

Regarding real-time: Sionna aims for fast prototyping and is not designed to be a real-time library. However, due to GPU acceleration, the achievable throughput can be quite competitive in many practical scenarios. This really depends on your exact configuration and requirements.
